nixos/modules/tasks/network-interfaces-systemd.nix

```diff
@@ -44,47 +44,53 @@ let
+    in
+    mkIf networkCfg.useDHCP {
+      networks."99-ethernet-default-dhcp" = {
+        # We want to match physical ethernet interfaces as commonly
+        # found on laptops, desktops and servers, to provide an
+        # "out-of-the-box" setup that works for common cases.  This
+        # heuristic isn't perfect (it could match interfaces with
+        # custom names that _happen_ to start with en or eth), but
+        # should be good enough to make the common case easy and can
+        # be overridden on a case-by-case basis using
+        # higher-priority networks or by disabling useDHCP.
 
-      # Type=ether matches veth interfaces as well, and this is
-      # more likely to result in interfaces being configured to
-      # use DHCP when they shouldn't.
+        # Type=ether matches veth interfaces as well, and this is
+        # more likely to result in interfaces being configured to
+        # use DHCP when they shouldn't.
 
-      # When wait-online.anyInterface is enabled, RequiredForOnline really
-      # means "sufficient for online", so we can enable it.
-      # Otherwise, don't block the network coming online because of default networks.
-      matchConfig.Name = ["en*" "eth*"];
-      DHCP = "yes";
-      linkConfig.RequiredForOnline =
-        lib.mkDefault (if initrd
-        then config.boot.initrd.systemd.network.wait-online.anyInterface
-        else config.systemd.network.wait-online.anyInterface);
-      networkConfig.IPv6PrivacyExtensions = "kernel";
-    };
-    networks."99-wireless-client-dhcp" = {
-      # Like above, but this is much more likely to be correct.
-      matchConfig.WLANInterfaceType = "station";
-      DHCP = "yes";
-      linkConfig.RequiredForOnline =
-        lib.mkDefault config.systemd.network.wait-online.anyInterface;
-      networkConfig.IPv6PrivacyExtensions = "kernel";
-      # We also set the route metric to one more than the default
-      # of 1024, so that Ethernet is preferred if both are
-      # available.
-      dhcpV4Config.RouteMetric = 1025;
-      ipv6AcceptRAConfig.RouteMetric = 1025;
+        # When wait-online.anyInterface is enabled, RequiredForOnline really
+        # means "sufficient for online", so we can enable it.
+        # Otherwise, don't block the network coming online because of default networks.
+        matchConfig.Name = [ "en*" "eth*" ];
+        DHCP = "yes";
+        linkConfig.RequiredForOnline =
+          lib.mkDefault (if initrd
+          then config.boot.initrd.systemd.network.wait-online.anyInterface
+          else config.systemd.network.wait-online.anyInterface);
+        networkConfig.IPv6PrivacyExtensions = "kernel";
+      };
+      networks."99-wireless-client-dhcp" = {
+        # Like above, but this is much more likely to be correct.
+        matchConfig.WLANInterfaceType = "station";
+        DHCP = "yes";
+        linkConfig.RequiredForOnline =
+          lib.mkDefault config.systemd.network.wait-online.anyInterface;
```

Comment:
While you're here, it seems like a bug that this uses `config.systemd.network.wait-online.anyInterface` instead of `networkCfg.wait-online.anyInterface`.